src/dataflow/src/source/mod.rs

```diff
@@ -1664,110 +1707,88 @@ impl SourceReaderPersistence {
         }
     }
 
-    /// Restores the known source partitions (along with their latest read offset, if the binding
-    /// is not beyond the `upper_data_seal_ts`) and the known timestamp bindings.
-    fn restore(
-        &self,
-    ) -> Result<
-        (
-            HashMap<PartitionId, Option<MzOffset>>,
-            Vec<(SourceTimestamp, AssignedTimestamp)>,
-        ),
-        persist::error::Error,
-    > {
-        let (offsets, bindings) = self.internal_restore(
-            self.config.read_handle.clone(),
-            self.config.upper_bindings_seal_ts,
-            self.config.upper_data_seal_ts,
-        )?;
-
-        tracing::trace!(
-            "In {}, initial (restored) source offsets: {:?}",
-            self.source_name,
-            offsets,
-        );
-
-        tracing::trace!(
-            "In {}, initial (restored) timestamp bindings: {:?}",
-            self.source_name,
-            bindings,
-        );
-
-        Ok((offsets, bindings))
-    }
-
-    /// Restores the latest partition offsets and timestamp bindings from the given `StreamReadHandle`.
+    /// Restores timestamp bindings from the given `StreamReadHandle` by reading the differential
+    /// timestamp updates and materializing/consolidating them into a `Vec`. This basically sums up
+    /// the `diff` of the bindings and collects those bindings whose `diff` is `1`.
+    ///
+    /// This restores two versions of the bindings: one as of an up to the common upper seal
```

Comment:
Ok this was challenging but I think I get the idea.

Is it basically -- if we've sealed at up to `cap_t3` and we have bindings

```
((offset1, t1), cap_t1, 1)
((offset2, t2), cap_t2, 1)
((offset3, t3), cap_t3, 1)
```

then we want to keep all bindings < `cap_t3` and so ... we first store in `TimestampBindingUpdater`

```
((offset1, t1), 1)
((offset2, t2), 1)
((offset3, t3), 1)
```

(which is latest_bindings). Then we store in timestamp histories

```
((offset1, t1), 1)
((offset2, t2), 1)
```

then we call update with timestamp histories

which negates what already exists in timestamp_updater and then adds what is in timestamp_histories

leaving us with `((offset3, cap_t3), -1)`? I believe that's the logic unless I grossly misunderstood something. I guess to me it seems like while we are in `restore` we already know that `((offset3, t3)` can be thrown out?

Like, this isn't a full diff but a sketch of what I'm thinking about


```
diff --git a/src/dataflow/src/source/mod.rs b/src/dataflow/src/source/mod.rs
index 7e7a4b450..a9fc05de5 100644
--- a/src/dataflow/src/source/mod.rs
+++ b/src/dataflow/src/source/mod.rs
@@ -1734,7 +1734,7 @@ impl SourceReaderPersistence {
         assert!(self.config.upper_bindings_seal_ts >= self.config.upper_data_seal_ts);

         // Materialized version of all bindings updates.
-        let mut latest_bindings: HashMap<_, isize> = HashMap::new();
+        let mut invalid_bindings: HashMap<_, isize> = HashMap::new();

         // Materialized version of bindings updates that are not beyond the common seal timestamp.
         let mut valid_bindings: HashMap<_, isize> = HashMap::new();
@@ -1751,13 +1751,11 @@ impl SourceReaderPersistence {
                 *valid_bindings
                     .entry((source_timestamp.clone(), assigned_timestamp))
                     .or_default() += diff;
+            } else {
+                *invalid_bindings
+                    .entry((source_timestamp, assigned_timestamp))
+                    .or_default() += diff;
             }
-
-            // We consolidate all the diffs as we go. Below we collect all positive updates and
-            // return them.
-            *latest_bindings
-                .entry((source_timestamp, assigned_timestamp))
-                .or_default() += diff;
         }
```

It seems like we have to jump through all of the hoops of
- seed the updater with the state we want to throw away + the state we want to keep
- seed the history with the state we want to keep
- call update which does gives us back (updater - history) = -(state we want to throw away)
- persist that state.

in order to make this work with the updater / timestamp histories API. I think this could be simplified a bit if
- restore returned a set of bindings to keep and a disjoint set of bindings to throw away
- we exposed a function that was something like `emit_bindings` or something like that which `maybe_emit_bindings` also calls?
- We only keep the state we want to keep in `updater` and `timestamp_histories`

It's late and I might be missing something fundamental about why that doesn't work. if so, im really sorry!

Reply:
The analysis of what it does is exactly correct! I will change it to not rely on the differentializer to do the retracts, and instead return them from the code that restores them, because there we know exactly what to retract.